File: src/utils/pestRiskRules.ts
import type { ForecastDay, NewAlert } from '../types/rules.ts'

export interface PestRiskInput {
  fieldId: string
  fieldName: string
  pointId?: number
  forecast: ForecastDay[]
  ndvi: number
  ndviFieldAvg: number
  crop: string
  growthStage: string
  recentAiAlertCount: number
}

export interface PestRiskResult {
  riskLevel: 'low' | 'medium' | 'high'
  factors: string[]
  window: string
  draftAlert?: NewAlert
}

function dayMeanTemp(day: ForecastDay): number {
  return (Number(day.tempMax) + Number(day.tempMin)) / 2
}

function hasHumid3d(days: ForecastDay[]): boolean {
  const sorted = [...days].sort((a, b) => a.date.localeCompare(b.date))
  for (let i = 0; i <= sorted.length - 3; i++) {
    const window = sorted.slice(i, i + 3)
    if (window.every((d) => Number(d.humidity ?? 0) > 80)) return true
  }
  return false
}

function rain7d(days: ForecastDay[]): number {
  const sorted = [...days].sort((a, b) => a.date.localeCompare(b.date)).slice(0, 7)
  return sorted.reduce((sum, d) => sum + Number(d.precipMm || 0), 0)
}

function meanTemp5d(days: ForecastDay[]): number | null {
  const sorted = [...days].sort((a, b) => a.date.localeCompare(b.date)).slice(0, 5)
  if (sorted.length === 0) return null
  const sum = sorted.reduce((acc, d) => acc + dayMeanTemp(d), 0)
  return sum / sorted.length
}

function scoreToLevel(score: number): PestRiskResult['riskLevel'] {
  if (score >= 4) return 'high'
  if (score >= 2) return 'medium'
  return 'low'
}

export function evaluatePestRisk(input: PestRiskInput): PestRiskResult {
  const factors: string[] = []
  const forecast = input.forecast || []

  if (hasHumid3d(forecast)) {
    factors.push('连续 3 日湿度 > 80%')
  }
  if (rain7d(forecast) > 80) {
    factors.push('7 日累计降水偏多')
  }
  if (input.ndviFieldAvg > 0 && input.ndvi < input.ndviFieldAvg * 0.85) {
    factors.push('NDVI 低于田间均值 15%')
  }
  const mean5 = meanTemp5d(forecast)
  if (
    mean5 != null &&
    mean5 >= 22 &&
    mean5 <= 28 &&
    String(input.crop).includes('小麦')
  ) {
    factors.push('气温处于病害流行适温区间')
  }
  if (input.recentAiAlertCount >= 2) {
    factors.push('近期 AI 已多次检出病虫害')
  }

  const riskLevel = scoreToLevel(factors.length)
  const window = forecast.length
    ? `${forecast[0].date}~${forecast[forecast.length - 1].date}`
    : ''

  const result: PestRiskResult = { riskLevel, factors, window }
  if (riskLevel === 'high') {
    result.draftAlert = {
      pointId: input.pointId ?? 0,
      fieldId: input.fieldId,
      level: 'high',
      message: `[虫情风险] 地块 ${input.fieldName} - 风险等级：high（${factors.join('；')}）`,
      time: Date.now(),
      handled: false,
      source: 'auto',
      ruleId: 'pest_risk',
      chain: 'pest',
      draft: true
    }
  }
  return result
}

File: ml-bjj/serving/rules/pest_risk_rules.py
from __future__ import annotations

from datetime import datetime


def _day_mean_temp(day: dict) -> float:
    return (float(day.get("tempMax") or 0) + float(day.get("tempMin") or 0)) / 2


def _has_humid_3d(days: list[dict]) -> bool:
    sorted_days = sorted(days, key=lambda item: str(item.get("date") or ""))
    for index in range(0, max(0, len(sorted_days) - 2)):
        window = sorted_days[index : index + 3]
        if all(float(item.get("humidity") or 0) > 80 for item in window):
            return True
    return False


def _rain_7d(days: list[dict]) -> float:
    sorted_days = sorted(days, key=lambda item: str(item.get("date") or ""))[:7]
    return sum(float(item.get("precipMm") or 0) for item in sorted_days)


def _mean_temp_5d(days: list[dict]) -> float | None:
    sorted_days = sorted(days, key=lambda item: str(item.get("date") or ""))[:5]
    if not sorted_days:
        return None
    return sum(_day_mean_temp(item) for item in sorted_days) / len(sorted_days)


def evaluate_pest_risk(input_data: dict) -> dict:
    factors: list[str] = []
    forecast = list(input_data.get("forecast") or [])
    if _has_humid_3d(forecast):
        factors.append("连续 3 日湿度 > 80%")
    if _rain_7d(forecast) > 80:
        factors.append("7 日累计降水偏多")
    ndvi = float(input_data.get("ndvi") or 0)
    ndvi_avg = float(input_data.get("ndviFieldAvg") or 0)
    if ndvi_avg > 0 and ndvi < ndvi_avg * 0.85:
        factors.append("NDVI 低于田间均值 15%")
    mean5 = _mean_temp_5d(forecast)
    crop = str(input_data.get("crop") or "")
    if mean5 is not None and 22 <= mean5 <= 28 and "小麦" in crop:
        factors.append("气温处于病害流行适温区间")
    if int(input_data.get("recentAiAlertCount") or 0) >= 2:
        factors.append("近期 AI 已多次检出病虫害")

    score = len(factors)
    risk_level = "high" if score >= 4 else "medium" if score >= 2 else "low"
    window = ""
    if forecast:
        first = str(forecast[0].get("date") or "")
        last = str(forecast[-1].get("date") or "")
        window = f"{first}~{last}"

    result: dict = {"riskLevel": risk_level, "factors": factors, "window": window}
    if risk_level == "high":
        field_name = input_data.get("fieldName") or input_data.get("fieldId")
        result["draftAlert"] = {
            "pointId": int(input_data.get("pointId") or 0),
            "fieldId": input_data.get("fieldId"),
            "level": "high",
            "message": f"[虫情风险] 地块 {field_name} - 风险等级：high（{'；'.join(factors)}）",
            "time": int(datetime.now().timestamp() * 1000),
            "handled": False,
            "source": "auto",
            "ruleId": "pest_risk",
            "chain": "pest",
            "draft": True,
        }
    return result
